### PR TITLE
LPS-117628, LPS-117641 Nested operation exception using DefaultWorkflowEngineImp

### PR DESCRIPTION
**Pull request by ctang3**
https://issues.liferay.com/browse/LPS-117628
https://issues.liferay.com/browse/LPS-117641

### CODE REVIEW

**Comment by ctang3** [comment]
ci:reevaluate:1387401_1639

File: modules/apps/portal-workflow/portal-workflow-kaleo-runtime-impl/src/main/java/com/liferay/portal/workflow/kaleo/runtime/internal/helper/NotificationHelperImpl.java [hunk 1]
```diff
@@ -47,6 +48,10 @@ public void sendKaleoNotifications(
 			ExecutionType executionType, ExecutionContext executionContext)
 		throws PortalException {
 
+		if (!CTCollectionThreadLocal.isProductionMode()) {
```
Comment (david-truong) [inline modules/apps/portal-workflow/portal-workflow-kaleo-runtime-impl/src/main/java/com/liferay/portal/workflow/kaleo/runtime/internal/helper/NotificationHelperImpl.java:51]: @ctang3 is there a technical reason for this or just a design decision?

**Review by david-truong — COMMENTED** [review]
- LPS-117628 Don't send workflow notifications in a publication - is there a technical reason we want to do this?
- LPS-117628 Use locale instead of resource bundle - I think we should consider doing this change in separate commit by itself

**Comment by ctang3** [comment]
ci:test

**Comment by ctang3** [comment]
@david-truong dropped those commits you mentioned

**Comment by ctang3** [comment]
ci:test:publications

**Comment by ctang3** [comment]
ci:test:relevant

**Comment by ctang3** [comment]
ci:test

**Comment by ctang3** [comment]
ci:test

**Comment by ctang3** [comment]
@xbrianlee could you take a look at these test results? thanks!

**Comment by xbrianlee** [comment]
ci:test:publications

**Comment by xbrianlee** [comment]
ci:test:relevant

**Comment by xbrianlee** [comment]
ci:test:sf

**Comment by xbrianlee** [comment]
@ctang3 Still going through these test results but I did notice a unique Workflow unit test failure so far
```     [exec] /opt/dev/projects/github/liferay-portal/modules/apps/portal-workflow/portal-workflow-kaleo-runtime-integration-impl/src/test/java/com/liferay/portal/workflow/kaleo/runtime/integration/internal/util/LazyWorkflowTaskAssigneeListTest.java:58: error: cannot access UnsafeFunction
     [exec] 				new KaleoTaskAssignmentInstanceLocalServiceWrapper() {
     [exec] 				                                                     ^
     [exec]   class file for com.liferay.petra.function.UnsafeFunction not found
     [exec] 1 error
```
([console log](https://testray.liferay.com/reports/production/logs/2022-07/test-1-38/test-portal-acceptance-pullrequest(master)/2527/modules-unit-jdk8/0/3/jenkins-console.txt.gz?authuser=0))

**Comment by xbrianlee** [comment]
ci:test:workflow

**Comment by xbrianlee** [comment]
ci:test:publications-environments

**Comment by suilin** [comment]
ci:test:workflow:
Some of the integration tests failed due to Compilation failed, such as https://testray.liferay.com/reports/production/logs/2022-07/test-1-32/test-portal-acceptance-pullrequest(master)/4942/modules-unit-jdk8/0/0/jenkins-console.txt.gz?authuser=0
`     [exec] * What went wrong:
     [exec] Execution failed for task ':apps:portal-workflow:portal-workflow-kaleo-runtime-integration-impl:compileTestJava'.
     [exec] > Compilation failed; see the compiler error output for details.
     [exec] 
     [exec] * Try:
     [exec] Run with --info or --debug option to get more log output. Run with --scan to get full insights.
     [exec] 
     [exec] * Exception is:
     [exec] org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':apps:portal-workflow:portal-workflow-kaleo-runtime-integration-impl:compileTestJava'.
     [exec] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:208)`

ci:test:publications-environments:
There are three functioanl tests on DB2 11.1 are unique, all of them are failed to login with new user on DB2
* Publications#AssertNonPrivilegedUserCannotSeePublicationsBar
* Publications#AllowPublicationsPermissionUserToAddAndPublishPublications
* Publications#CanResetPasswordWhenLastInPublication

Steps to reproduce:
* Enable Publications
* Add a new user with Publications User
* Add a new publication
* Logout and login with the new user

Result: Fail to login and following [errors](https://testray.liferay.com/reports/production/logs/2022-07/test-1-38/test-portal-acceptance-pullrequest(master)/2573/functional-tomcat90-db2111-jdk8/0/1/liferay-log.txt.gz?authuser=0) throws.

**Comment by xbrianlee** [comment]
The same portal-workflow-kaleo-runtime-integration-impl compile error is present in this latest workflow test suite run as well. Recent test runs don't seem to have this error so I believe it's related to the changes in this PR.

**Comment by ctang3** [comment]
@xbrianlee @suilin  I'll take a look at this, thank you both!

**Comment by ctang3** [comment]
ci:test:workflow

**Comment by xbrianlee** [comment]
There's a [CI issue](https://liferay.slack.com/archives/CL3TWSMHQ/p1658947803893909) currently that's causing the error above

**Comment by xbrianlee** [comment]
ci:test:workflow

**Comment by xbrianlee** [comment]
ci:test:publications

**Comment by xbrianlee** [comment]
ci:test:relevant

**Comment by suilin** [comment]
ci:test:publications-environments

**Comment by suilin** [comment]
Hi @ctang3, the issues still can be reproduced on the latest commits, can you look at it? Thanks.
* Workflow integration tests still have the same failures:
**   Cannot access UnsafeFunction, [here](https://testray.liferay.com/reports/production/logs/2022-07/test-1-35/test-portal-acceptance-pullrequest(master)/2150/modules-unit-jdk8/0/0/jenkins-console.txt.gz?authuser=0) is the details

- [ ]      [exec] /opt/dev/projects/github/liferay-portal/modules/apps/portal-workflow/portal-workflow-kaleo-runtime-integration-impl/src/test/java/com/liferay/portal/workflow/kaleo/runtime/integration/internal/util/LazyWorkflowTaskAssigneeListTest.java:58: error: cannot access UnsafeFunction
- [ ]      [exec] 				new KaleoTaskAssignmentInstanceLocalServiceWrapper() {
- [ ]      [exec] 				                                                     ^
- [ ]      [exec]   class file for com.liferay.petra.function.UnsafeFunction not found

** apps:portal-workflow:portal-workflow-kaleo-runtime-integration-impl:compile error, details on [here](https://testray.liferay.com/reports/production/logs/2022-07/test-1-36/test-portal-acceptance-pullrequest(master)/1684/modules-unit-jdk8/0/0/jenkins-console.txt.gz?authuser=0).
* semantic-versioning issue for com.liferay.change.tracking.spi.display

**Comment by ctang3** [comment]
ci:test:publications-environments

**Comment by ctang3** [comment]
Rebased and pushed fix commits.

@suilin #1 and #2 from your comment here should be fixed by https://github.com/liferay-publications/liferay-portal/pull/122/commits/5f4decd5a96a0720f2601bf7df8c146cc535c145 and https://github.com/liferay-publications/liferay-portal/pull/122/commits/ef3d55fb0cc2ce5fdb68cf677b18655edef44e41. I am not able to reproduce #3 on my local after rebasing.

**Comment by suilin** [comment]
ci:test:workflow

**Comment by suilin** [comment]
Hi @ctang3, thanks for fixing them, the results look good, can we forward it?

**Comment by ctang3** [comment]
@suilin yes, I'll go ahead and do that.

**Comment by ctang3** [comment]
ci:forward

**Comment by suilin** [comment]
ci:forward

**Comment by xbrianlee** [comment]
Fixed the SF error and manually forwarded to https://github.com/brianchandotcom/liferay-portal/pull/121262

**Pull request closed without merging**